src/kernels/jupyter/jupyterRemoteCachedKernelValidator.ts

```diff
@@ -0,0 +1,68 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import { inject, injectable } from 'inversify';
+import { traceWarning } from '../../platform/logging';
+import { LiveRemoteKernelConnectionMetadata } from '../types';
+import { computeServerId, extractJupyterServerHandleAndId } from './jupyterUtils';
+import {
+    IJupyterRemoteCachedKernelValidator,
+    IJupyterServerUriStorage,
+    IJupyterUriProviderRegistration,
+    ILiveRemoteKernelConnectionUsageTracker
+} from './types';
+
+@injectable()
+export class JupyterRemoteCachedKernelValidator implements IJupyterRemoteCachedKernelValidator {
+    constructor(
+        @inject(ILiveRemoteKernelConnectionUsageTracker)
+        private readonly liveKernelConnectionTracker: ILiveRemoteKernelConnectionUsageTracker,
+
+        @inject(IJupyterServerUriStorage) private readonly uriStorage: IJupyterServerUriStorage,
+        @inject(IJupyterUriProviderRegistration) private readonly providerRegistration: IJupyterUriProviderRegistration
+    ) {}
+    public async isValid(kernel: LiveRemoteKernelConnectionMetadata): Promise<boolean> {
+        // Only list live kernels that was used by the user,
+        if (!this.liveKernelConnectionTracker.wasKernelUsed(kernel)) {
+            return false;
+        }
+        const providersPromise = this.providerRegistration.getProviders();
+        const currentList = await this.uriStorage.getSavedUriList();
+        const item = currentList.find((item) => computeServerId(item.uri) === kernel.serverId);
+        if (!item) {
+            // Server has been removed and we have some old cached data.
+            return false;
+        }
+        // Check if this has a provider associated with it.
+        const info = extractJupyterServerHandleAndId(item.uri);
+        if (!info) {
+            // Could be a regular remote Jupyter Uri entered by the user.
+            // As its in the list, its still valid.
+            return true;
+        }
+        const providers = await providersPromise;
+        const provider = providers.find((item) => item.id === info.id);
+        if (!provider) {
+            // Extension may have been uninstalled.
```

Comment:
After a couple of readings, now I wonder: would logs make sense instead of comments?